File: tests/performance/benchmark.php
<?php
declare(strict_types=1);

require_once __DIR__ . '/BenchmarkSummary.php';

$options = getopt('', [
    'wordpress-dir:',
    'base-url:',
    'samples::',
    'warmups::',
    'output:',
    'summary:',
    'git-sha::',
]);

$wordpressDir = isset($options['wordpress-dir']) ? rtrim((string) $options['wordpress-dir'], '/\\') : '';
$baseUrl      = isset($options['base-url']) ? rtrim((string) $options['base-url'], '/') : '';
$outputPath   = isset($options['output']) ? (string) $options['output'] : '';
$summaryPath  = isset($options['summary']) ? (string) $options['summary'] : '';
$samples      = isset($options['samples']) ? (int) $options['samples'] : 15;
$warmups      = isset($options['warmups']) ? (int) $options['warmups'] : 1;
$gitSha       = isset($options['git-sha']) ? (string) $options['git-sha'] : '';

if ( '' === $wordpressDir || '' === $baseUrl || '' === $outputPath || '' === $summaryPath ) {
    fwrite(STDERR, "Required: --wordpress-dir --base-url --output --summary\n");
    exit(2);
}
if ( $samples < 5 ) {
    fwrite(STDERR, "At least five measured pairs are required.\n");
    exit(2);
}
if ( $warmups < 1 ) {
    fwrite(STDERR, "At least one warm-up request is required before each measured state sample.\n");
    exit(2);
}
if ( ! is_dir($wordpressDir) ) {
    fwrite(STDERR, "WordPress directory does not exist: {$wordpressDir}\n");
    exit(2);
}
if ( ! function_exists('curl_init') ) {
    fwrite(STDERR, "PHP cURL extension is required.\n");
    exit(2);
}

$outputDir = dirname($outputPath);
$summaryDir = dirname($summaryPath);
foreach ( array_unique([$outputDir, $summaryDir]) as $dir ) {
    if ( ! is_dir($dir) && ! mkdir($dir, 0777, true) && ! is_dir($dir) ) {
        throw new RuntimeException("Unable to create output directory {$dir}.");
    }
}

$probeDir = $wordpressDir . '/wp-content/wddtf-benchmark-output';
if ( ! is_dir($probeDir) && ! mkdir($probeDir, 0777, true) && ! is_dir($probeDir) ) {
    throw new RuntimeException('Unable to create benchmark probe output directory.');
}

$cookieJar = tempnam(sys_get_temp_dir(), 'wddtf-perf-cookie-');
if ( false === $cookieJar ) {
    throw new RuntimeException('Unable to create benchmark cookie jar.');
}

register_shutdown_function(static function() use ($cookieJar): void {
    if ( is_file($cookieJar) ) {
        @unlink($cookieJar);
    }
});

/** @return string */
function runWp(string $wordpressDir, array $arguments): string {
    $parts = ['wp', '--path=' . $wordpressDir];
    foreach ( $arguments as $argument ) {
        $parts[] = (string) $argument;
    }

    $command = implode(' ', array_map('escapeshellarg', $parts)) . ' 2>&1';
    exec($command, $lines, $exitCode);
    $output = trim(implode("\n", $lines));

    if ( 0 !== $exitCode ) {
        throw new RuntimeException("WP-CLI failed ({$exitCode}): {$command}\n{$output}");
    }

    return $output;
}

function setDeepState(string $wordpressDir, string $state): void {
    if ( 'active' === $state ) {
        runWp($wordpressDir, ['plugin', 'activate', 'wp-deep-diagnostics', '--quiet']);
        return;
    }

    runWp($wordpressDir, ['plugin', 'deactivate', 'wp-deep-diagnostics', '--quiet']);
}

function cleanProductReports(string $wordpressDir): void {
    $dir = $wordpressDir . '/wp-content/uploads/wp-deep-diagnostics';
    foreach ( glob($dir . '/report-*') ?: [] as $file ) {
        if ( is_file($file) && ! @unlink($file) ) {
            throw new RuntimeException("Unable to remove prior benchmark report {$file}.");
        }
    }
}

/** @return array{body:string,http_code:int,effective_url:string,wall_ms:float} */
function httpRequest(string $url, ?string $cookieJar, ?array $postFields = null): array {
    $handle = curl_init($url);
    if ( false === $handle ) {
        throw new RuntimeException('Unable to initialize cURL.');
    }

    $curlOptions = [
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_FOLLOWLOCATION => true,
        CURLOPT_MAXREDIRS      => 5,
        CURLOPT_CONNECTTIMEOUT => 5,
        CURLOPT_TIMEOUT        => 30,
        CURLOPT_USERAGENT      => 'WDDTF-Performance-Qualification/1',
    ];

    if ( null !== $cookieJar && '' !== $cookieJar ) {
        $curlOptions[CURLOPT_COOKIEJAR]  = $cookieJar;
        $curlOptions[CURLOPT_COOKIEFILE] = $cookieJar;
    }

    curl_setopt_array($handle, $curlOptions);

    if ( null !== $postFields ) {
        curl_setopt($handle, CURLOPT_POST, true);
        curl_setopt($handle, CURLOPT_POSTFIELDS, http_build_query($postFields, '', '&', PHP_QUERY_RFC3986));
    }

    $started = hrtime(true);
    $body = curl_exec($handle);
    $ended = hrtime(true);

    if ( false === $body ) {
        $error = curl_error($handle);
        curl_close($handle);
        throw new RuntimeException("HTTP request failed for {$url}: {$error}");
    }

    $result = [
        'body'          => (string) $body,
        'http_code'     => (int) curl_getinfo($handle, CURLINFO_RESPONSE_CODE),
        'effective_url' => (string) curl_getinfo($handle, CURLINFO_EFFECTIVE_URL),
        'wall_ms'       => round(($ended - $started) / 1_000_000, 3),
    ];
    curl_close($handle);

    return $result;
}

function loginAdmin(string $baseUrl, string $cookieJar): void {
    $loginUrl = $baseUrl . '/wp-login.php';
    $landing = httpRequest($loginUrl, $cookieJar);
    if ( 200 !== $landing['http_code'] ) {
        throw new RuntimeException('Unable to load WordPress login page.');
    }

    $response = httpRequest($loginUrl, $cookieJar, [
        'log'         => 'admin',
        'pwd'         => 'ci-password',
        'wp-submit'   => 'Log In',
        'redirect_to' => $baseUrl . '/wp-admin/',
        'testcookie'  => '1',
    ]);

    if ( 200 !== $response['http_code'] || str_contains($response['effective_url'], 'wp-login.php') ) {
        throw new RuntimeException('Authenticated wp-admin session could not be established.');
    }
}

/** @return array<string,mixed> */
function measuredRequest(
    string $baseUrl,
    string $path,
    ?string $cookieJar,
    string $probeDir,
    string $sampleId,
    bool $expectActive,
    bool $requiresAdmin
): array {
    $separator = str_contains($path, '?') ? '&' : '?';
    $url = $baseUrl . $path . $separator . 'wddtf_benchmark_sample=' . rawurlencode($sampleId);
    $metricPath = $probeDir . '/' . $sampleId . '.json';
    @unlink($metricPath);

    $response = httpRequest($url, $cookieJar);
    clearstatcache(true, $metricPath);

    if ( 200 !== $response['http_code'] ) {
        throw new RuntimeException("Benchmark request {$sampleId} returned HTTP {$response['http_code']}.");
    }
    if ( $requiresAdmin && str_contains($response['effective_url'], 'wp-login.php') ) {
        throw new RuntimeException("Benchmark request {$sampleId} lost authenticated wp-admin context.");
    }

    // Validity boundary: the post-finalize benchmark probe runs at shutdown/PHP_INT_MAX.
    // If its file is not already visible when the external client returns, client timing
    // cannot truthfully be claimed to include late shutdown/finalization work.
    if ( ! is_file($metricPath) ) {
        throw new RuntimeException(
            "Benchmark request {$sampleId} returned before the post-shutdown marker was observable; wall time is invalid."
        );
    }

    $decoded = json_decode((string) file_get_contents($metricPath), true, 512, JSON_THROW_ON_ERROR);
    if ( ! is_array($decoded) || ($decoded['sample_id'] ?? null) !== $sampleId ) {
        throw new RuntimeException("Benchmark probe output for {$sampleId} is malformed.");
    }
    if ( ! array_key_exists('deep_active', $decoded) || (bool) $decoded['deep_active'] !== $expectActive ) {
        throw new RuntimeException("Benchmark request {$sampleId} did not run in the expected plugin state.");
    }
    if ( ! is_int($decoded['peak_memory_bytes'] ?? null) || ! is_int($decoded['db_query_count'] ?? null) ) {
        throw new RuntimeException("Benchmark request {$sampleId} is missing memory/query evidence.");
    }
    if ( ! is_float($decoded['late_shutdown_ms'] ?? null) && ! is_int($decoded['late_shutdown_ms'] ?? null) ) {
        throw new RuntimeException("Benchmark request {$sampleId} is missing late-shutdown evidence.");
    }

    return [
        'wall_ms'           => $response['wall_ms'],
        'peak_memory_bytes' => $decoded['peak_memory_bytes'],
        'db_query_count'    => $decoded['db_query_count'],
        'late_shutdown_ms'  => (float) $decoded['late_shutdown_ms'],
    ];
}

function formatMs(float|int $value): string {
    return number_format((float) $value, 2, '.', '');
}

function formatMiB(float|int $bytes): string {
    return number_format(((float) $bytes) / 1048576, 2, '.', '');
}

try {
    setDeepState($wordpressDir, 'control');
    loginAdmin($baseUrl, $cookieJar);

    $scenarios = [
        'frontend' => [
            'label' => 'Normal anonymous frontend',
            'path' => '/',
            'requires_admin' => false,
        ],
        'admin' => [
            'label' => 'Normal authenticated wp-admin',
            'path' => '/wp-admin/index.php',
            'requires_admin' => true,
        ],
        'diagnostic_workload' => [
            'label' => 'Bounded diagnostic workload',
            'path' => '/wp-admin/tools.php?page=wddtf-performance-fixture',
            'requires_admin' => true,
        ],
    ];

    $rawSamples = [];

    foreach ( $scenarios as $scenarioId => $scenario ) {
        for ( $pair = 1; $pair <= $samples; $pair++ ) {
            $order = 1 === ($pair % 2) ? ['control', 'active'] : ['active', 'control'];

            foreach ( $order as $state ) {
                setDeepState($wordpressDir, $state);
                cleanProductReports($wordpressDir);
                $requestCookieJar = $scenario['requires_admin'] ? $cookieJar : null;

                for ( $warmup = 1; $warmup <= $warmups; $warmup++ ) {
                    $warmupId = sprintf('warmup-%s-p%d-%s-%d', $scenarioId, $pair, $state, $warmup);
                    measuredRequest(
                        $baseUrl,
                        $scenario['path'],
                        $requestCookieJar,
                        $probeDir,
                        $warmupId,
                        'active' === $state,
                        $scenario['requires_admin']
                    );
                }

                // Keep report-retention history out of the sample itself. The directory is
                // already warm after the warm-up; only prior report-* files are removed.
                cleanProductReports($wordpressDir);

                $sampleId = sprintf('sample-%s-p%d-%s', $scenarioId, $pair, $state);
                $metrics = measuredRequest(
                    $baseUrl,
                    $scenario['path'],
                    $requestCookieJar,
                    $probeDir,
                    $sampleId,
                    'active' === $state,
                    $scenario['requires_admin']
                );

                $rawSamples[] = [
                    'scenario'          => $scenarioId,
                    'state'             => $state,
                    'pair'              => $pair,
                    'wall_ms'           => $metrics['wall_ms'],
                    'peak_memory_bytes' => $metrics['peak_memory_bytes'],
                    'db_query_count'    => $metrics['db_query_count'],
                    'late_shutdown_ms'  => $metrics['late_shutdown_ms'],
                ];
            }
        }
    }

    $summaries = WddtfPerformanceBenchmarkSummary::summarize($rawSamples, $samples);

    $wordpressVersion = runWp($wordpressDir, ['core', 'version']);
    $pluginVersion = runWp($wordpressDir, ['plugin', 'get', 'wp-deep-diagnostics', '--field=version']);
    $databaseVersion = runWp($wordpressDir, ['eval', 'global $wpdb; echo $wpdb->db_version();']);
    $saveQueries = '1' === runWp($wordpressDir, ['eval', 'echo defined("SAVEQUERIES") && SAVEQUERIES ? "1" : "0";']);
    $httpBlocked = '1' === runWp($wordpressDir, ['eval', 'echo defined("WP_HTTP_BLOCK_EXTERNAL") && WP_HTTP_BLOCK_EXTERNAL ? "1" : "0";']);

    $scenarioOutput = [];
    foreach ( $scenarios as $scenarioId => $scenario ) {
        $scenarioSamples = array_values(array_filter(
            $rawSamples,
            static fn(array $sample): bool => $sample['scenario'] === $scenarioId
        ));

        $scenarioOutput[] = [
            'id'      => $scenarioId,
            'label'   => $scenario['label'],
            'auth'    => $scenario['requires_admin'] ? 'authenticated-admin' : 'anonymous',
            'samples' => $scenarioSamples,
            'summary' => $summaries[$scenarioId],
        ];
    }

    $result = [
        'schema_version' => 1,
        'generated_at_utc' => gmdate('c'),
        'environment' => [
            'git_sha' => $gitSha,
            'deep_diagnostics_version' => $pluginVersion,
            'wordpress_version' => $wordpressVersion,
            'php_version' => PHP_VERSION,
            'database_version' => $databaseVersion,
            'server_sapi' => 'cli-server',
            'opcache_enable_cli' => (bool) ini_get('opcache.enable_cli'),
            'savequeries' => $saveQueries,
            'external_http_blocked' => $httpBlocked,
        ],
        'measurement' => [
            'measured_pairs_per_scenario' => $samples,
            'warmups_before_each_state_sample' => $warmups,
            'pair_order' => 'odd pairs control-active; even pairs active-control',
            'wall_clock' => 'external PHP cURL client monotonic time; validity requires post-shutdown probe marker to exist before client completion',
            'late_shutdown' => 'shutdown action priority 9998 through PHP_INT_MAX; active path includes Manager::finalize() at shutdown priority 9999 plus any later shutdown work',
            'memory' => 'PHP request peak real memory at post-shutdown probe',
            'database_queries' => 'wpdb->num_queries at post-shutdown probe; does not require SAVEQUERIES',
            'report_history_policy' => 'report-* files removed outside measured requests after warm-up so samples measure a warm report directory without accumulating benchmark history',
        ],
        'scenarios' => $scenarioOutput,
        'caveats' => [
            'No numeric performance acceptance threshold is enforced in this characterization batch.',
            'Disposable WordPress/GitHub-hosted-runner evidence is not universal production overhead proof.',
            'SAVEQUERIES is not enabled in the canonical baseline; its generic WordPress cost is therefore not attributed to Deep Diagnostics.',
            'The late-shutdown metric is a narrow phase boundary around Deep finalization, not a generic profiler trace.',
        ],
    ];

    $encoded = json_encode($result, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES | JSON_THROW_ON_ERROR) . "\n";
    file_put_contents($outputPath, $encoded, LOCK_EX);

    $markdown = [];
    $markdown[] = '# Deep Diagnostics performance/overhead characterization';
    $markdown[] = '';
    $markdown[] = sprintf(
        'Environment: WordPress %s, PHP %s, database %s, Deep Diagnostics %s. Samples: %d matched pairs/scenario; %d warm-up request(s) before every measured state sample.',
        $wordpressVersion,
        PHP_VERSION,
        $databaseVersion,
        $pluginVersion,
        $samples,
        $warmups
    );
    $markdown[] = '';
    $markdown[] = '| Scenario | Wall control median (range), ms | Wall active median (range), ms | Paired wall delta median [p25–p75], ms | Peak-memory paired delta median, MiB | DB-query paired delta median | Late-shutdown paired delta median, ms |';
    $markdown[] = '| --- | ---: | ---: | ---: | ---: | ---: | ---: |';

    foreach ( $scenarioOutput as $scenario ) {
        $summary = $scenario['summary'];
        $wall = $summary['wall_ms'];
        $memory = $summary['peak_memory_bytes'];
        $queries = $summary['db_query_count'];
        $shutdown = $summary['late_shutdown_ms'];

        $markdown[] = sprintf(
            '| %s | %s (%s–%s) | %s (%s–%s) | %s [%s–%s] | %s | %s | %s |',
            $scenario['label'],
            formatMs($wall['control']['median']),
            formatMs($wall['control']['min']),
            formatMs($wall['control']['max']),
            formatMs($wall['active']['median']),
            formatMs($wall['active']['min']),
            formatMs($wall['active']['max']),
            formatMs($wall['delta']['paired']['median']),
            formatMs($wall['delta']['paired']['p25']),
            formatMs($wall['delta']['paired']['p75']),
            formatMiB($memory['delta']['paired']['median']),
            formatMs($queries['delta']['paired']['median']),
            formatMs($shutdown['delta']['paired']['median'])
        );
    }

    $markdown[] = '';
    $markdown[] = 'Positive deltas mean the active request used more time/memory/queries than its matched inactive control. Negative wall-clock pair deltas can occur from runner noise; use the median and distribution rather than a single request.';
    $markdown[] = '';
    $markdown[] = '**Validity policy:** this job fails for missing pairs, malformed metrics, plugin-state mismatch, missing sample counts, or when the external client returns before the post-shutdown marker is observable. It intentionally does **not** fail on any arbitrary overhead number.';
    $markdown[] = '';
    $markdown[] = '**Claim ceiling:** these numbers characterize this disposable runtime only. They do not establish production traffic, arbitrary hosting, every plugin/theme combination, or SRWF production behavior.';
    $markdown[] = '';

    file_put_contents($summaryPath, implode("\n", $markdown), LOCK_EX);

    echo "Benchmark result: {$outputPath}\n";
    echo "Benchmark summary: {$summaryPath}\n";
} catch (Throwable $error) {
    fwrite(STDERR, "Performance qualification failed: {$error->getMessage()}\n");
    exit(1);
}

File: tests/performance/benchmark.py
#!/usr/bin/env python3
"""Paired external overhead benchmark for WP Deep Diagnostics."""
from __future__ import annotations

import argparse
import http.cookiejar
import json
import os
import statistics
import subprocess
import sys
import time
import urllib.parse
import urllib.request
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

SCHEMA_VERSION = 1


def run(*args: str, cwd: Path | None = None) -> str:
    completed = subprocess.run(
        list(args),
        cwd=str(cwd) if cwd else None,
        check=True,
        text=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
    return completed.stdout.strip()


def percentile(values: list[float], fraction: float) -> float:
    ordered = sorted(values)
    if not ordered:
        raise ValueError("cannot summarize empty sample set")
    if len(ordered) == 1:
        return ordered[0]
    position = (len(ordered) - 1) * fraction
    lower = int(position)
    upper = min(lower + 1, len(ordered) - 1)
    weight = position - lower
    return ordered[lower] * (1 - weight) + ordered[upper] * weight


def summarize(values: list[float]) -> dict[str, float]:
    return {
        "median": round(statistics.median(values), 3),
        "min": round(min(values), 3),
        "p25": round(percentile(values, 0.25), 3),
        "p75": round(percentile(values, 0.75), 3),
        "max": round(max(values), 3),
    }


def load_records(path: Path) -> list[dict[str, Any]]:
    if not path.exists():
        return []
    records: list[dict[str, Any]] = []
    for line in path.read_text(encoding="utf-8").splitlines():
        if line.strip():
            records.append(json.loads(line))
    return records


def wait_for_record(path: Path, token: str, deadline_s: float = 5.0) -> tuple[dict[str, Any], float]:
    started = time.perf_counter_ns()
    deadline = time.monotonic() + deadline_s
    while time.monotonic() < deadline:
        for record in reversed(load_records(path)):
            if record.get("token") == token:
                waited_ms = (time.perf_counter_ns() - started) / 1_000_000
                return record, waited_ms
        time.sleep(0.005)
    raise RuntimeError(f"probe record not observed for token {token}")


def request(opener: urllib.request.OpenerDirector, url: str) -> tuple[int, float]:
    started = time.perf_counter_ns()
    with opener.open(url, timeout=15) as response:
        status = int(response.status)
        response.read()
    elapsed_ms = (time.perf_counter_ns() - started) / 1_000_000
    return status, elapsed_ms


def login(base_url: str, username: str, password: str) -> urllib.request.OpenerDirector:
    jar = http.cookiejar.CookieJar()
    opener = urllib.request.build_opener(urllib.request.HTTPCookieProcessor(jar))
    opener.open(f"{base_url}/wp-login.php", timeout=15).read()
    payload = urllib.parse.urlencode(
        {
            "log": username,
            "pwd": password,
            "wp-submit": "Log In",
            "redirect_to": f"{base_url}/wp-admin/",
            "testcookie": "1",
        }
    ).encode()
    request_obj = urllib.request.Request(f"{base_url}/wp-login.php", data=payload)
    with opener.open(request_obj, timeout=15) as response:
        response.read()
        if "/wp-admin" not in response.geturl():
            raise RuntimeError("benchmark admin login did not reach wp-admin")
    return opener


def ensure_plugin_state(wp_path: Path, active: bool) -> None:
    completed = subprocess.run(
        ["wp", "plugin", "is-active", "wp-deep-diagnostics", "--path", str(wp_path)],
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
        text=True,
    )
    if (completed.returncode == 0) != active:
        if active:
            run("wp", "plugin", "activate", "wp-deep-diagnostics", "--quiet", "--path", str(wp_path))
        else:
            run("wp", "plugin", "deactivate", "wp-deep-diagnostics", "--quiet", "--path", str(wp_path))


def metric_values(samples: list[dict[str, Any]], key: str) -> list[float]:
    values: list[float] = []
    for sample in samples:
        value = sample.get(key)
        if isinstance(value, (int, float)):
            values.append(float(value))
    return values


def summarize_samples(samples: list[dict[str, Any]]) -> dict[str, Any]:
    fields = [
        "client_wall_ms",
        "completion_wall_ms",
        "post_response_wait_ms",
        "peak_memory_bytes",
        "memory_usage_bytes",
        "query_count",
        "plugins_loaded_window_ms",
        "shutdown_window_ms",
    ]
    summary: dict[str, Any] = {"sample_count": len(samples), "metrics": {}}
    for field in fields:
        values = metric_values(samples, field)
        summary["metrics"][field] = summarize(values) if len(values) == len(samples) else None
    return summary


def build_url(base_url: str, path: str, scenario: str, token: str) -> str:
    separator = "&" if "?" in path else "?"
    query = urllib.parse.urlencode({"wddtf_bench_scenario": scenario, "wddtf_bench": token})
    return f"{base_url}{path}{separator}{query}"


def run_state(
    *,
    state: str,
    active: bool,
    scenario_id: str,
    path: str,
    opener: urllib.request.OpenerDirector,
    base_url: str,
    metrics_file: Path,
    wp_path: Path,
    warmups: int,
    samples: int,
) -> dict[str, Any]:
    ensure_plugin_state(wp_path, active)
    metrics_file.write_text("", encoding="utf-8")

    for index in range(warmups):
        token = f"warmup-{scenario_id}-{state}-{index}"
        status, _ = request(opener, build_url(base_url, path, scenario_id, token))
        if status != 200:
            raise RuntimeError(f"warm-up returned HTTP {status} for {scenario_id}/{state}")
        wait_for_record(metrics_file, token)

    metrics_file.write_text("", encoding="utf-8")
    measured: list[dict[str, Any]] = []

    for index in range(samples):
        token = f"sample-{scenario_id}-{state}-{index}"
        request_started = time.perf_counter_ns()
        status, client_wall_ms = request(opener, build_url(base_url, path, scenario_id, token))
        record, post_wait_ms = wait_for_record(metrics_file, token)
        completion_wall_ms = (time.perf_counter_ns() - request_started) / 1_000_000

        if status != 200:
            raise RuntimeError(f"HTTP {status} for {scenario_id}/{state}")
        if record.get("scenario") != scenario_id:
            raise RuntimeError(f"probe scenario mismatch for {token}")
        if bool(record.get("deep_active")) != active:
            raise RuntimeError(f"plugin-state mismatch for {token}")

        measured.append(
            {
                "index": index,
                "client_wall_ms": round(client_wall_ms, 3),
                "completion_wall_ms": round(completion_wall_ms, 3),
                "post_response_wait_ms": round(post_wait_ms, 3),
                "peak_memory_bytes": record.get("peak_memory_bytes"),
                "memory_usage_bytes": record.get("memory_usage_bytes"),
                "query_count": record.get("query_count"),
                "plugins_loaded_window_ms": round(float(record["plugins_loaded_window_ms"]), 3)
                if isinstance(record.get("plugins_loaded_window_ms"), (int, float))
                else None,
                "shutdown_window_ms": round(float(record["shutdown_window_ms"]), 3)
                if isinstance(record.get("shutdown_window_ms"), (int, float))
                else None,
            }
        )

    return {"plugin_state": state, "samples": measured, "summary": summarize_samples(measured)}


def median_metric(state: dict[str, Any], metric: str) -> float | None:
    block = state["summary"]["metrics"].get(metric)
    return float(block["median"]) if isinstance(block, dict) else None


def render_summary(result: dict[str, Any]) -> str:
    lines = [
        "# Deep Diagnostics overhead qualification",
        "",
        f"Environment: WordPress {result['environment']['wordpress']}, PHP {result['environment']['php']}, DB {result['environment']['database']}",
        f"Warm-ups: {result['policy']['warmup_requests_per_state']} per state; measured samples: {result['policy']['measured_requests_per_state']} per state.",
        "",
        "| Scenario | Control median completion | Active median completion | Median delta | Peak-memory delta | Query-count delta | Active shutdown window |",
        "| --- | ---: | ---: | ---: | ---: | ---: | ---: |",
    ]
    for scenario in result["scenarios"]:
        delta = scenario["delta_active_minus_control"]
        control = scenario["states"]["inactive"]
        active = scenario["states"]["active"]
        lines.append(
            "| {name} | {c:.3f} ms | {a:.3f} ms | {d:.3f} ms | {m:.0f} B | {q:.3f} | {s:.3f} ms |".format(
                name=scenario["id"],
                c=median_metric(control, "completion_wall_ms") or 0.0,
                a=median_metric(active, "completion_wall_ms") or 0.0,
                d=delta.get("completion_wall_ms") or 0.0,
                m=delta.get("peak_memory_bytes") or 0.0,
                q=delta.get("query_count") or 0.0,
                s=median_metric(active, "shutdown_window_ms") or 0.0,
            )
        )
    lines.extend(
        [
            "",
            "`completion_wall_ms` is the primary external/request-level timing. It runs from client request start until the test-only probe has observed the WordPress shutdown hook after Deep Diagnostics finalization.",
            "`client_wall_ms` is also retained. A non-trivial `post_response_wait_ms` would indicate that the client saw the response before the post-finalization probe record became visible.",
            "No numeric product budget is enforced by this workflow; only benchmark validity and result-contract integrity are blocking.",
            "",
        ]
    )
    return "\n".join(lines)


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--wp-path", type=Path, required=True)
    parser.add_argument("--base-url", required=True)
    parser.add_argument("--metrics-file", type=Path, required=True)
    parser.add_argument("--output", type=Path, required=True)
    parser.add_argument("--summary", type=Path, required=True)
    parser.add_argument("--admin-user", default="admin")
    parser.add_argument("--admin-password", default="ci-password")
    parser.add_argument("--page-id", required=True)
    parser.add_argument("--warmups", type=int, default=5)
    parser.add_argument("--samples", type=int, default=15)
    args = parser.parse_args()

    if args.warmups < 1 or args.samples < 5:
        raise SystemExit("benchmark requires at least 1 warm-up and 5 measured samples per state")

    anonymous = urllib.request.build_opener()
    authenticated = login(args.base_url.rstrip("/"), args.admin_user, args.admin_password)

    scenarios = [
        {"id": "frontend", "path": f"/?page_id={args.page_id}", "authenticated": False, "savequeries": False},
        {"id": "admin", "path": "/wp-admin/index.php", "authenticated": True, "savequeries": False},
        {"id": "diagnostic_workload", "path": "/wp-admin/index.php", "authenticated": True, "savequeries": False},
    ]

    result: dict[str, Any] = {
        "schema_version": SCHEMA_VERSION,
        "generated_at": datetime.now(timezone.utc).replace(microsecond=0).isoformat(),
        "environment": {
            "wordpress": run("wp", "core", "version", "--path", str(args.wp_path)),
            "php": run("wp", "eval", "echo PHP_VERSION;", "--path", str(args.wp_path)),
            "database": run("wp", "eval", 'global $wpdb; echo $wpdb->get_var("SELECT VERSION()");', "--path", str(args.wp_path)),
            "deep_diagnostics": run("wp", "plugin", "get", "wp-deep-diagnostics", "--field=version", "--path", str(args.wp_path)),
            "runner_os": os.environ.get("RUNNER_OS", "unknown"),
            "php_sapi": "cli-server",
        },
        "policy": {
            "warmup_requests_per_state": args.warmups,
            "measured_requests_per_state": args.samples,
            "state_order": ["inactive", "active"],
            "primary_metric": "completion_wall_ms",
            "savequeries_default": False,
            "hard_overhead_budget": None,
        },
        "scenarios": [],
        "caveats": [
            "Disposable single-runner WordPress/MySQL qualification; not production traffic evidence.",
            "The MU-plugin probe is test-only and runs in both control and active states.",
            "SAVEQUERIES is disabled in canonical scenarios, so its generic WordPress overhead is not attributed to Deep Diagnostics.",
            "Phase windows are controlled-runtime attribution aids, not a generic profiler.",
        ],
    }

    for scenario in scenarios:
        opener = authenticated if scenario["authenticated"] else anonymous
        states: dict[str, Any] = {}
        states["inactive"] = run_state(
            state="inactive",
            active=False,
            scenario_id=scenario["id"],
            path=scenario["path"],
            opener=opener,
            base_url=args.base_url.rstrip("/"),
            metrics_file=args.metrics_file,
            wp_path=args.wp_path,
            warmups=args.warmups,
            samples=args.samples,
        )
        states["active"] = run_state(
            state="active",
            active=True,
            scenario_id=scenario["id"],
            path=scenario["path"],
            opener=opener,
            base_url=args.base_url.rstrip("/"),
            metrics_file=args.metrics_file,
            wp_path=args.wp_path,
            warmups=args.warmups,
            samples=args.samples,
        )

        delta: dict[str, float | None] = {}
        for metric in [
            "client_wall_ms",
            "completion_wall_ms",
            "post_response_wait_ms",
            "peak_memory_bytes",
            "memory_usage_bytes",
            "query_count",
            "plugins_loaded_window_ms",
            "shutdown_window_ms",
        ]:
            active_median = median_metric(states["active"], metric)
            inactive_median = median_metric(states["inactive"], metric)
            delta[metric] = round(active_median - inactive_median, 3) if active_median is not None and inactive_median is not None else None

        result["scenarios"].append(
            {
                "id": scenario["id"],
                "request_path": scenario["path"],
                "authenticated": scenario["authenticated"],
                "savequeries": scenario["savequeries"],
                "states": states,
                "delta_active_minus_control": delta,
            }
        )

    if result["schema_version"] != 1 or len(result["scenarios"]) != 3:
        raise RuntimeError("invalid result contract")
    for scenario in result["scenarios"]:
        for state in ("inactive", "active"):
            block = scenario["states"][state]
            if block["summary"]["sample_count"] != args.samples or len(block["samples"]) != args.samples:
                raise RuntimeError(f"missing required samples for {scenario['id']}/{state}")
            for sample in block["samples"]:
                for required in ("completion_wall_ms", "peak_memory_bytes", "query_count"):
                    if not isinstance(sample.get(required), (int, float)):
                        raise RuntimeError(f"malformed {required} for {scenario['id']}/{state}")

    args.output.parent.mkdir(parents=True, exist_ok=True)
    args.output.write_text(json.dumps(result, indent=2, sort_keys=True) + "\n", encoding="utf-8")
    args.summary.write_text(render_summary(result), encoding="utf-8")
    print(render_summary(result))
    return 0


if __name__ == "__main__":
    sys.exit(main())
